### PR TITLE
fix: handle shell arguments on Windows

### LINKED ISSUE #192: Path(s) are added automatically to shell command on Windows
https://github.com/sxyazi/yazi/assets/33489972/63042c3e-c655-44b2-9982-93882f18f11b

This issue has been reported via Discord but I'll also leave it here to keep track.

### PR DESCRIPTION
**Pull request by ndtoan96**
Fix #192

### CODE REVIEW

**Comment by sxyazi** [comment]
I've made a simple test with:

```rust
cmdexpand::expand_cmd(r#"  a "b \"%1\""  "#, &["c", "d"])
```

And got:

```
  a "b \"
```

Is this expected?

**Comment by ndtoan96** [comment]
I'm aware that it has problem with escaped quote, but I spent yesterday fixing it in my crate and making documentation so haven't ported it here. Will find time to do it later.

**Comment by ndtoan96** [comment]
Just realized this expand thing only works with no-space argument. I tried to open a text file `"D:\a b.txt"` by select the file and run `%EDITOR% %1`. Even though it expands correctly to `%EDITOR% "D:\a b.txt"`, the editor actually open 2 files `"D:\a` and `b.txt"`.

Pretty sure it's because the way `cmd /C` works, argument should be feed one by one instead of all in one.

**Comment by sxyazi** [comment]
I have to sleep, will investigate it tomorrow. =)

**Comment by sxyazi** [comment]
I just tested it, and it indeed doesn't work as expected:

```bash
cmd /C "nvim 'rust fmt.toml'"
```

This results in opening two files (`'rust` and `fmt.toml'`), and `cmd` doesn't correctly recognize the quotes. I'm thinking of two simple workarounds now:

- Use PowerShell instead of cmd.
  I found that PowerShell can work correctly:

  ```bash
  powershell -Command "& nvim 'rust fmt.toml'"
  ```

  However, this requires PowerShell to be present on the user's system, it requires a minimum of Windows 8. In fact, I haven't tested whether Yazi works correctly on Windows 7 and earlier systems yet...

- Parse only the first-level placeholders, and, if they exist, pass them as cmd parameters:

  From:

  ```bash
  cmd /C "nvim 'rust fmt.toml'"
  ```

  To:

  ```rust
  Command::new("cmd").args(["/C", "nvim", "rust fmt.toml"])
  ```

 It was similar to Yazi's current behavior, but the difference is that this time we can use `nom` to parse more accurately which first-level placeholders are available.

Apart from these, I'm not sure if there are better solutions. Any ideas?

**Comment by ndtoan96** [comment]
I prefer the second approach. Reason is that user can use environment variable in the form `%SOME_VAR%`, which is nicer, and more well known than `$env:SOME_VAR` in powershell.

**Comment by sxyazi** [comment]
Oh, that makes sense, It's the first time I've seen this kind of peculiar format 🤣

**Comment by ndtoan96** [comment]
I thought it would be easy enough to make a manual parser. The code doesn't look pretty but it works (at least if users don't do anything too crazy).😅

File: core/src/external/shell.rs [hunk 1]
```diff
@@ -49,3 +50,166 @@ pub fn shell(opt: ShellOpt) -> Result<Child> {
...
+		#[test]
+		fn test_double_quote() {
+			let args = parse_cmd_to_args("echo \"abc ' 'xyz\" \"%1\" %2 %3", &["111", "222"]);
+			assert_eq!(args, vec!["echo", "abc ' 'xyz", "111", "222", ""]);
+		}
+
+		#[test]
+		fn test_escaped() {
+			let args = parse_cmd_to_args("echo \"a\tbc ' 'x\nyz\" \"\\%1\" %2 %3", &["111", "22  2"]);
+			assert_eq!(args, vec!["echo", "a\tbc ' 'x\nyz", "%1", "22  2", ""]);
```
Comment (sxyazi) [inline core/src/external/shell.rs:194]: This test case seems wrong; the fourth element should be `"\\%1"` instead of `"%1"`?

Comment (sxyazi) [reply core/src/external/shell.rs:194]: Got it!

**Comment by sxyazi** [comment]
I've made some simplifications, merging now. Thanks!

**Pull request merged**